### DISCUSSION TITLE
Multiple XML rows with same r attr

### DISCUSSION THREAD

**Discussion by thomascharbonnel** [qax-os/excelize#1308] [Q&A] [answered]
Hello,

I'm working on row styles right now but i'm facing an issue related to the XML parsing logic because both `Next()` and `Columns()` increment the XML element.
I've found a way to fix this by returning the `rowIterator.columns` when reaching a closing `</row>` tag, but there are two tests that are failing, they're similar in the way they work: https://github.com/qax-os/excelize/blob/master/cell_test.go#L246.
They're failing because the XML contains two rows with the same `r="2"` attribute.

I was wondering whether this is supported by the XSLX spec and if Excel actually returns XML files containing multiple rows with the same ID? Removing that test (if it is not necessary) would make my life quite a bit easier! Here is the change i'm suggesting: https://github.com/qax-os/excelize/pull/1297/commits/dd1fb8b7b9e0162c11a2372b3fa615e57704ce29

An alternative idea i had in mind was to just load all the columns when parsing the row in `Next()` and storing the columns in `Rows{}`, then `Columns()` would not contain any XML logic.
Another option (i think) would be to use a `TeeReader` and two XML decoders moving a different speed (though that's more complicated)

**Comment by xuri** [comment, marked as answer]
The test for multiple rows with same `r` attribute comes from real valid file, which can be supported by Excel, so we needn't remove it. I've left some comments on the pull request #1297 about this issue.
